### PR TITLE
Add more std::io documentation.

### PR DESCRIPTION
**Pull request by steveklabnik**
This round: io::Result and the free functions.

### CODE REVIEW

File: src/libstd/io/error.rs [hunk 1]
```diff
@@ -17,11 +17,35 @@ use option::Option::{self, Some, None};
...
+///
+/// This type is broadly used across `std::io` for any operation which may produce an error.
+///
+/// This typedef is generally used to avoid writing out `io::Error` directly and is otherwise a
+/// direct mapping to `Result`.
+///
+/// While usual Rust style is to import types directly, aliases of `Result` often are not, to make
+/// it easier to distinguish between them. `Result` is generally assumed to be
+/// `std::result::Result`, and so users of this alias will generally use `io::Result` instead of
+/// shadowing the prelud's import of `std::result::Result`.
```
Comment (alexcrichton) [inline src/libstd/io/error.rs]: prelude

File: src/libstd/io/stdio.rs [hunk 2]
```diff
@@ -154,15 +154,43 @@ pub struct StdinLock<'a> {
...
 }
 
-/// Creates a new handle to the global standard input stream of this process.
+/// Constructs a new handle to the standard input of the current process.
 ///
-/// The handle returned refers to a globally shared buffer between all threads.
-/// Access is synchronized and can be explicitly controlled with the `lock()`
-/// method.
+/// Each handle returned is a reference to a shared global buffer whose access is synchronized via
+/// a mutex. If you need more explicit control over locking, see the [lock() method][lock].
```
Comment (alexcrichton) [inline src/libstd/io/stdio.rs]: Could this (and the other docs here) stick to 80 chars? I believe all of `std::io` is already formatted this way and it's nice to just keep it consistent.

File: src/libstd/io/stdio.rs [hunk 3]
```diff
@@ -154,15 +154,43 @@ pub struct StdinLock<'a> {
...
-/// locked version, `StdinLock`, implements both `Read` and `BufRead`, however.
+/// [lock]: struct.Stdin.html#method.lock
+///
+/// # Examples
+///
+/// Using implicit synchronization:
+///
+/// ```
+/// use std::io;
+/// use std::io::Read;
```
Comment (alexcrichton) [inline src/libstd/io/stdio.rs]:
```rust
use std::io::{self, Read};
```

File: src/libstd/io/stdio.rs [hunk 4]
```diff
@@ -376,12 +434,43 @@ pub struct StderrLock<'a> {
     inner: ReentrantMutexGuard<'a, RefCell<Maybe<StderrRaw>>>,
 }
 
-/// Constructs a new reference to the standard error stream of a process.
+/// Constructs a new handle to the standard error of the current process.
+///
+/// Each handle returned is a reference to a shared global buffer whose access is synchronized via
+/// a mutex. If you need more explicit control over locking, see the [lock() method][lock].
```
Comment (alexcrichton) [inline src/libstd/io/stdio.rs]: This should continue to mention that there is no buffering I believe.

File: src/libstd/io/stdio.rs (hunk 3 above)
Comment (steveklabnik) [inline src/libstd/io/stdio.rs]: Is this the official style? I kinda hate it :/

File: src/libstd/io/stdio.rs (hunk 2 above)
Comment (steveklabnik) [inline src/libstd/io/stdio.rs]: /me hopes @nrc gets `rustfmt` up to snuff soon enough

**Comment by steveklabnik** [comment]
@alexcrichton updated!

File: src/libstd/io/error.rs [hunk 5]
```diff
@@ -17,11 +17,35 @@ use option::Option::{self, Some, None};
...
 
-/// A type for results generated by I/O related functions where the `Err` type
-/// is hard-wired to `io::Error`.
+/// A specialized [`Result`][result] type for I/O operations.
 ///
-/// This typedef is generally used to avoid writing out `io::Error` directly and
-/// is otherwise a direct mapping to `std::result::Result`.
+/// [result]: ../result/enum.Result.html
+///
+/// This type is broadly used across `std::io` for any operation which may produce an error.
```
Comment (alexcrichton) [inline src/libstd/io/error.rs]: (same 80-char concern here as I had below)

File: src/libstd/io/stdio.rs [hunk 6]
```diff
@@ -154,15 +154,43 @@ pub struct StdinLock<'a> {
...
+/// locking, see the [lock() method][lock].
+///
+/// [lock]: struct.Stdin.html#method.lock
+///
+/// # Examples
+///
+/// Using implicit synchronization:
+///
+/// ```
+/// use std::io{self, Read};
```
Comment (alexcrichton) [inline src/libstd/io/stdio.rs]: s/io{/io::{/

File: src/libstd/io/stdio.rs [hunk 7]
```diff
@@ -154,15 +154,43 @@ pub struct StdinLock<'a> {
...
+/// try!(io::stdin().read_to_string(&mut buffer));
+/// # Ok(buffer)
+/// # }
+/// ```
+///
+/// Using explicit synchronization:
+///
+/// ```
+/// use std::io;
+/// use std::io::Read;
```
Comment (alexcrichton) [inline src/libstd/io/stdio.rs]: Same `self` treatment here

Comment (alexcrichton) [inline src/libstd/io/stdio.rs]: (and in snippets below)

File: src/libstd/io/stdio.rs [hunk 8]
```diff
@@ -376,12 +435,46 @@ pub struct StderrLock<'a> {
     inner: ReentrantMutexGuard<'a, RefCell<Maybe<StderrRaw>>>,
 }
 
-/// Constructs a new reference to the standard error stream of a process.
+/// Constructs a new handle to the standard error of the current process.
+///
+/// Each handle returned is a reference to a shared global buffer whose access
```
Comment (alexcrichton) [inline src/libstd/io/stdio.rs]: The "shared global buffer" here may want to be dropped

File: src/libstd/io/util.rs [hunk 9]
```diff
@@ -28,6 +28,22 @@ use io::{self, Read, Write, ErrorKind, BufRead};
...
+/// ```
+/// use std::io;
+///
+/// # fn foo() -> io::Result<()> {
+/// let mut reader = &(b"hello")[..];
+/// let mut writer: Vec<u8> = vec![];
+///
+/// try!(io::copy(&mut reader, &mut writer));
+///
+/// assert_eq!(&reader[..], &writer[..]);
```
Comment (alexcrichton) [inline src/libstd/io/util.rs]: There's quite a bit of syntax going on here! I feel like this assert shouldn't need the slicing syntax, and the `&(b"...")[...]` above may want to be replaced with type ascription:

```rust
let mut reader: &[u8] = b"hello";
```

File: src/libstd/io/util.rs [hunk 10]
```diff
@@ -48,9 +64,24 @@ pub fn copy<R: Read, W: Write>(reader: &mut R, writer: &mut W) -> io::Result<u64
...
 pub struct Empty { _priv: () }
 
-/// Creates an instance of an empty reader.
+/// Constructs a new handle to an empty reader.
 ///
 /// All reads from the returned reader will return `Ok(0)`.
+///
+/// # Examples
+///
+/// Reading a bunch of zeroes, for fun:
```
Comment (alexcrichton) [inline src/libstd/io/util.rs]: This doesn't actually read any zeros, but instead doesn't read anything (e.g. doesn't modify the buffer passed in)

File: src/libstd/io/error.rs (hunk 5 above)
Comment (steveklabnik) [inline src/libstd/io/error.rs]: ugh, i specifically tried to watch out for this for you :(

File: src/libstd/io/stdio.rs (hunk 8 above)
Comment (steveklabnik) [inline src/libstd/io/stdio.rs]: lol, my bad "this is buffered" next line "this is not buffered"

File: src/libstd/io/util.rs (hunk 9 above)
Comment (steveklabnik) [inline src/libstd/io/util.rs]: without the slicing, it doesn't work :/

File: src/libstd/io/util.rs (hunk 10 above)
Comment (steveklabnik) [inline src/libstd/io/util.rs]: ahhh i mis-read `All reads from the returned reader will return Ok(0)` as meaning that it just read zeroes, but that's the EOF thing :(

File: src/libstd/io/stdio.rs (hunk 6 above)
Comment (steveklabnik) [inline src/libstd/io/stdio.rs]: :sweat:

File: src/libstd/io/util.rs (hunk 9 above)
Comment (steveklabnik) [inline src/libstd/io/util.rs]: for posterity:

```
error: the trait `core::cmp::PartialEq<collections::vec::Vec<u8>>` is not implemented for the type `&[u8]` [E0277]
```

i guess i can eliminate the slice of the second, though.

**Comment by steveklabnik** [comment]
@alexcrichton updated again :)

File: src/rust-installer [hunk 11]
```diff
@@ -1 +1 @@
-Subproject commit 8e4f8ea581502a2edc8177a040300e05ff7f91e3
+Subproject commit 0da191a30ba385215c5eb1dc97c2b5f076f93b07
```
Comment (alexcrichton) [inline src/rust-installer]: accidental update?

**Comment by alexcrichton** [comment]
Looks good to me, thanks @steveklabnik!

r=me

**Pull request merged**